test/suites/storage/suite_test.go

```diff
@@ -236,6 +239,142 @@ var _ = Describe("Persistent Volumes", func() {
 	})
 })
 
+var _ = Describe("Stateful workloads", func() {
+	var numPods int
+	var persistentVolumeClaim *corev1.PersistentVolumeClaim
+	var storageClass *storagev1.StorageClass
+	var statefulSet *appsv1.StatefulSet
+	var selector labels.Selector
+	BeforeEach(func() {
+		// Ensure that the EBS driver is installed, or we can't run the test.
+		var ds appsv1.DaemonSet
+		if err := env.Client.Get(env.Context, client.ObjectKey{
+			Namespace: "kube-system",
+			Name:      "ebs-csi-node",
+		}, &ds); err != nil {
+			if errors.IsNotFound(err) {
+				Skip(fmt.Sprintf("skipping StatefulSet test due to missing EBS driver %s", err))
+			} else {
+				Fail(fmt.Sprintf("determining EBS driver status, %s", err))
+			}
+		}
+
+		numPods = 1
+		subnets := env.GetSubnets(map[string]string{"karpenter.sh/discovery": env.ClusterName})
+		shuffledAZs := lo.Shuffle(lo.Keys(subnets))
+
+		storageClass = test.StorageClass(test.StorageClassOptions{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "test-storage-class",
+			},
+			Provisioner:       awssdk.String("ebs.csi.aws.com"),
+			VolumeBindingMode: lo.ToPtr(storagev1.VolumeBindingWaitForFirstConsumer),
+		})
+
+		storageClass.AllowedTopologies = []corev1.TopologySelectorTerm{{
```

Comment:
FYI added this to **ensure** sure the replacement node that spins up has same zone as the volume. 

I would have thought Karpenter already ensures this by checking for node volume topology of Unschedulable + Disruptable pods in the Provisioning controller before nominating pod, but something I saw in some test logs makes me not so sure that this functionality works in karpenter. 

This addition decreases the scope of this test from "Make sure no 6+ attach minute delays AND that replacement nodes respect disrupted stateful pod's volume topology" to just "make sure no 6+ min force attach delay", which is why it's worth adding.

Reply:
Another TODO post v1 comment:

Do you all have a disruption e2e test that checks that the replacement nodes for disrupted pods WITH PVCs respect volume topology when `volumeBindingMode: WaitForFirstConsumer`? 

I see that you have ["should run a pod with a dynamic persistent volume while respecting allowed topologies"](https://github.com/aws/karpenter-provider-aws/blob/a41bad8531f39f6bae566c87caadd6d69fec42f4/test/suites/storage/suite_test.go#L175), but that does not test for the disrupted -> replacement node case

I'm asking because at some point I saw the following pod events during a test:

```
time=2024-07-31T15:21:18Z type=Normal reason=Nominated from=karpenter message=Pod should schedule on: nodeclaim/carpetsun-73-u1srk6xvco-5bbzc
time=2024-07-31T15:22:27Z type=Normal reason=Scheduled from=default-scheduler message=Successfully assigned default/thundercloud-77-i6tboadfye-0 to ip-192-168-159-146.us-east-2.compute.internal
time=2024-07-31T15:22:31Z type=Normal reason=SuccessfulAttachVolume from=attachdetach-controller message=AttachVolume.Attach succeeded for volume "pvc-61c2ddee-172a-4047-851a-5984c3198d0b"
time=2024-07-31T15:22:35Z type=Normal reason=Created from=kubelet message=Created container crushernavy-79-wg3pt8vwq3
time=2024-07-31T15:22:35Z type=Normal reason=Started from=kubelet message=Started container crushernavy-79-wg3pt8vwq3
time=2024-07-31T15:22:37Z type=Normal reason=Evicted from=karpenter message=Evicted pod
time=2024-07-31T15:22:38Z type=Normal reason=Killing from=kubelet message=Stopping container crushernavy-79-wg3pt8vwq3
time=2024-07-31T15:22:39Z type=Warning reason=FailedScheduling from=default-scheduler message=0/3 nodes are available: 1 node(s) had untolerated taint {karpenter.sh/disrupted: }, 2 node(s) had untolerated taint {CriticalAddonsOnly: true}. preemption: 0/3 nodes are available: 3 Preemption is not helpful for scheduling.
time=2024-07-31T15:22:40Z type=Normal reason=Nominated from=karpenter message=Pod should schedule on: nodeclaim/carpetsun-73-u1srk6xvco-kxchq
time=2024-07-31T15:23:19Z type=Warning reason=FailedScheduling from=default-scheduler message=0/3 nodes are available: 1 node(s) had volume node affinity conflict, 2 node(s) had untolerated taint {CriticalAddonsOnly: true}. preemption: 0/3 nodes are available: 3 Preemption is not helpful for scheduling.
time=2024-07-31T15:23:33Z type=Normal reason=Scheduled from=default-scheduler message=Successfully assigned default/thundercloud-77-i6tboadfye-0 to ip-192-168-143-20.us-east-2.compute.internal
```

Note how long it took pod from killing of pod to finally being scheduled (55 sec)

Also note that there was a FailedScheduling event with `1 node(s) had volume node affinity conflict` 

Note, I don't think kubernetes gives you this feature for free, and this issue likely occurs with  other auto-scalers. Happy to work with ya'll on this feature after v1 releases 😅

Reply:
Could not reproduce issue, resolving.